### DISCUSSION TITLE
How to use a variable before it was declared in the code without executing that code at that time

### DISCUSSION THREAD

**Discussion by finkelshtein** [quarto-dev/quarto-cli#10859] [Q&A] [answered]
### Description

Hi,

I am working on a temporal solution for the problem described in Discussion #10844, using some find/replace and R code chunks. Most of the things are working, however, I have cases where I need to make a "reference" (not a Quarto reference, a self-made one) to a theorem-like environment which is somewhere below. I tried to use `ref.label` option. Here is the minimal (and simplified) example:

````
```{r}
#| echo: false
cur = 24
```

```{r, ref.label="ex1", echo = FALSE}
```

We consider this in Example `{r} ex1_label` below.

... some text ... 

```{r ex1}
#| echo: false
cur = cur + 1
ex1_label = cur
```

The value of `cur` now is `{r} cur`
````

If you run this code, the value of `cur` will be 26. I need to keep it equal to 25. (Because I want to use it to assign the correct current number in the code for that example.)

My understanding is that the code `cur = cur + 1` is executed in the chunk where `ref.label` is written. Is a way to get the value of `ex1_label` without executing the corresponding code too early?

**Comment by finkelshtein** [comment]
To be clear: the current output of the code above is 
![image](https://github.com/user-attachments/assets/d32ca6a5-8a0a-4246-bb19-7ca924020307)
I would like to have 25 in both cases.

**Reply by mcanouil** [reply, marked as answer]
That's not how it works.
All code are executed sequentially as it would in regular script.
If you change the variable, then you overwrite the previous value, that standard programming behaviour and is unrelated to Quarto.
You need to change your programming logic here.

Also, do not mix syntaxes for options.
Stick to one. In Quarto, the recommended syntax is YAML.

````qmd
---
title: "Quarto Playground"
format: html
---

```{r}
#| echo: false
cur = 24
```

```{r}
#| echo: false
#| ref-label: ex1
```

We consider this in Example `{r} ex1_label` below.

... some text ... 

```{r}
#| label: ex1
#| echo: false
cur = cur + 1
ex1_label = cur
```

The value of `cur` now is `{r} cur`
````
